python/ray/serve/batching.py

```diff
@@ -97,52 +96,41 @@ def __init__(
 
     def put(self, request: Tuple[_SingleRequest, asyncio.Future]) -> None:
         self.queue.put_nowait(request)
-        # Signal when the full batch is ready. The event will be reset
-        # in wait_for_batch.
-        if self.queue.qsize() == self.max_batch_size:
-            self.full_batch_event.set()
 
     async def wait_for_batch(self) -> List[Any]:
         """Wait for batch respecting self.max_batch_size and self.timeout_s.
 
-        Returns a batch of up to self.max_batch_size items, waiting for up
-        to self.timeout_s for a full batch. After the timeout, returns as many
-        items as are ready.
+        Returns a batch of up to self.max_batch_size items. Waits for up to
+        to self.timeout_s after receiving the first request that will be in
+        the next batch. After the timeout, returns as many items as are ready.
 
         Always returns a batch with at least one item - will block
         indefinitely until an item comes in.
         """
-        curr_timeout = self.timeout_s
+
         batch = []
-        while len(batch) == 0:
-            loop_start = time.time()
-
-            # If the timeout is 0, wait for any item to be available on the
-            # queue.
-            if curr_timeout == 0:
-                batch.append(await self.queue.get())
-            # If the timeout is nonzero, wait for either the timeout to occur
-            # or the max batch size to be ready.
-            else:
-                try:
-                    await asyncio.wait_for(self.full_batch_event.wait(), curr_timeout)
-                except asyncio.TimeoutError:
-                    pass
-
-            # Pull up to the max_batch_size requests off the queue.
-            while len(batch) < self.max_batch_size and not self.queue.empty():
-                batch.append(self.queue.get_nowait())
-
-            # Reset the event if there are fewer than max_batch_size requests
-            # in the queue.
-            if (
-                self.queue.qsize() < self.max_batch_size
-                and self.full_batch_event.is_set()
-            ):
-                self.full_batch_event.clear()
-
-            # Adjust the timeout based on the time spent in this iteration.
-            curr_timeout = max(0, curr_timeout - (time.time() - loop_start))
+        batch.append(await self.queue.get())
+
+        # Wait self.timeout_s seconds for new queue arrivals.
+        batch_start_time = time.time()
+        while (
+            time.time() - batch_start_time < self.timeout_s
+            and len(batch) < self.max_batch_size
+        ):
+            remaining_batch_time_s = max(
+                self.timeout_s - (time.time() - batch_start_time), 0
+            )
+            try:
+                batch.append(
+                    await asyncio.wait_for(self.queue.get(), remaining_batch_time_s)
```

Comment:
[minor]

you can optimize this a little bit by using `while not self.queue.empty()` and `self.queue.get_nowait()` to avoid unnecessarily yielding when there are results ready immediately

Reply:
I added that in the `while` loop afterwards. I felt it was a bit more readable to avoid nested `while` loops here.